### PR TITLE
[Docs] Describe building without network access

### PR DESCRIPTION
**Pull request by woju**
Signed-off-by: Wojtek Porczyk <woju@invisiblethingslab.com>

Fixes: gramineproject/graphene#1846

## How to test this PR?

Reread the instructions and do them manually.

### CODE REVIEW

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1018)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @mkow and @woju)

___
*[`Documentation/devel/building.rst` line 321 at r1](https://reviewable.io/reviews/gramineproject/gramine/1018#-NFh9pzl0XEU-oBdIqZi:-NFh9pzm7r20doG1OagQ:b-ph8a5g) ([raw file](https://github.com/gramineproject/gramine/blob/0921338caa2e28e321cc5851e6c820e3730d9015/Documentation/devel/building.rst#L321)):*
> ```reStructuredText
> First, before you cut your network access, you need to download (or otherwise
> obtain) a |~| checkout of Gramine repository and all wrapped subprojects'
> distfiles. The files :file:`subprojects/{*}.wrap` directory describe those
> ```

Why do you say `directory`? These are just files.
___
*[`Documentation/devel/building.rst` line 330 at r1](https://reviewable.io/reviews/gramineproject/gramine/1018#-NFhAGyLDbOIvL38NZgN:-NFhAGyLDbOIvL38NZgO:bg7z8nq) ([raw file](https://github.com/gramineproject/gramine/blob/0921338caa2e28e321cc5851e6c820e3730d9015/Documentation/devel/building.rst#L330)):*
> ```reStructuredText
> Alternatively, you can prepare a |~| "dist" tarball, which apart from Gramine
> code will contain all wrapped subprojects and also git submodules. You need to
> create a |~| dummy builddir for this, the chosen options don't matter as long as
> ```

`the chosen options` -- you mean for `meson setup`? Could you add it explicitly?

Maybe this:
```
For this, you need to create a |~| dummy builddir via :command:`meson setup` plus
:command:`meson dist`; the chosen config options don't matter as long as ...
```
___
*[`Documentation/devel/building.rst` line 346 at r1](https://reviewable.io/reviews/gramineproject/gramine/1018#-NFhBpV87nNR1KdASimf:-NFhBpV9BJE3JThqfKBP:b-6dlf0v) ([raw file](https://github.com/gramineproject/gramine/blob/0921338caa2e28e321cc5851e6c820e3730d9015/Documentation/devel/building.rst#L346)):*
> ```reStructuredText
> You can now sever your network connection::
> 
>     sudo unshare -n su "$USER"
> ```

Personal preference: `--net` instead of `-n`
___
*[`Documentation/devel/building.rst` line 353 at r1](https://reviewable.io/reviews/gramineproject/gramine/1018#-NFhBO567wuMAGLYlBCe:-NFhBO567wuMAGLYlBCf:bmdz59c) ([raw file](https://github.com/gramineproject/gramine/blob/0921338caa2e28e321cc5851e6c820e3730d9015/Documentation/devel/building.rst#L353)):*
> ```reStructuredText
> setup` your build directory with the switch ``--wrap-mode=nodownload``, which
> prevents Meson from downloading subprojects. Those subprojects should already be
> downloaded and if you didn't :command:`unshare -n`, it prevents a |~| mistake.
> ```

I can't parse this "it prevents a mistake". Which mistake? The mistake that the user is still having internet access, because he forgot to `unshare --net`?

Maybe rephrase it to something like "Without `--wrap-mode=nodownload`, Meson will try to download the projects and <fail?>".

**Review by mkow — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1018)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @woju)

___
*[`Documentation/devel/building.rst` line 324 at r1](https://reviewable.io/reviews/gramineproject/gramine/1018#-NFoCdEE4kFeJXGOnXP5:-NFoCdEE4kFeJXGOnXP6:b-pq5xd4) ([raw file](https://github.com/gramineproject/gramine/blob/0921338caa2e28e321cc5851e6c820e3730d9015/Documentation/devel/building.rst#L324)):*
> Otherwise, you should put all those distfiles into :file:`subprojects/packagecache` directory.

There's one problem with this description: it's missing how you should name these files, and it's not the naming which `wget` will use when downloading them.
Except from this it works for me.

**Review by dimakuv — COMMENTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1018)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @woju)

**Review by woju — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1018)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @mkow)

___
*[`Documentation/devel/building.rst` line 321 at r1](https://reviewable.io/reviews/gramineproject/gramine/1018#-NFh9pzl0XEU-oBdIqZi:-NFs7Mp63UsWeXgdrTPV:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/0921338caa2e28e321cc5851e6c820e3730d9015/Documentation/devel/building.rst#L321)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Why do you say `directory`? These are just files.
</blockquote></details>

Done.
___
*[`Documentation/devel/building.rst` line 324 at r1](https://reviewable.io/reviews/gramineproject/gramine/1018#-NFoCdEE4kFeJXGOnXP5:-NFs6f_G5az9cXAVgf7g:b-lkl3j1) ([raw file](https://github.com/gramineproject/gramine/blob/0921338caa2e28e321cc5851e6c820e3730d9015/Documentation/devel/building.rst#L324)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

> Otherwise, you should put all those distfiles into :file:`subprojects/packagecache` directory.

There's one problem with this description: it's missing how you should name these files, and it's not the naming which `wget` will use when downloading them.
Except from this it works for me.
</blockquote></details>

With `wget --content-disposition` and `curl -O` it's OK I think and I didn't want to describe the wrap file format (it's in `source_filename` key).
___
*[`Documentation/devel/building.rst` line 330 at r1](https://reviewable.io/reviews/gramineproject/gramine/1018#-NFhAGyLDbOIvL38NZgN:-NFsB1Cs05LXh-BTqlxw:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/0921338caa2e28e321cc5851e6c820e3730d9015/Documentation/devel/building.rst#L330)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`the chosen options` -- you mean for `meson setup`? Could you add it explicitly?

Maybe this:
```
For this, you need to create a |~| dummy builddir via :command:`meson setup` plus
:command:`meson dist`; the chosen config options don't matter as long as ...
```
</blockquote></details>

Done.
___
*[`Documentation/devel/building.rst` line 346 at r1](https://reviewable.io/reviews/gramineproject/gramine/1018#-NFhBpV87nNR1KdASimf:-NFs3QTY5jGc1O2PvN8w:b3uzgf) ([raw file](https://github.com/gramineproject/gramine/blob/0921338caa2e28e321cc5851e6c820e3730d9015/Documentation/devel/building.rst#L346)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Personal preference: `--net` instead of `-n`
</blockquote></details>

`-n` is shorter...
___
*[`Documentation/devel/building.rst` line 353 at r1](https://reviewable.io/reviews/gramineproject/gramine/1018#-NFhBO567wuMAGLYlBCe:-NFs3Rt3Dw6UaqGDPEVW:bdx1bgd) ([raw file](https://github.com/gramineproject/gramine/blob/0921338caa2e28e321cc5851e6c820e3730d9015/Documentation/devel/building.rst#L353)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

I can't parse this "it prevents a mistake". Which mistake? The mistake that the user is still having internet access, because he forgot to `unshare --net`?

Maybe rephrase it to something like "Without `--wrap-mode=nodownload`, Meson will try to download the projects and <fail?>".
</blockquote></details>

If you have all distfiles cached meson won't try to download, so you may be lucky. The mistake is that user *both* fails to unshare *and* forgets to download one or more distfiles. The problem is that, you may try it once and it works (won't download anything), so you do it more and get complacent. Then we update submodules, user pulls and oops, s/he just connected to the network.

**Review by dimakuv — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1018)* status: all files reviewed, 2 unresolved discussions, "fixup! " found in commit messages' one-liners (waiting on @mkow)

___
*[`Documentation/devel/building.rst` line 353 at r1](https://reviewable.io/reviews/gramineproject/gramine/1018#-NFhBO567wuMAGLYlBCe:-NFxLihcBSMdpoYflSMM:b-cda9d) ([raw file](https://github.com/gramineproject/gramine/blob/0921338caa2e28e321cc5851e6c820e3730d9015/Documentation/devel/building.rst#L353)):*
<details><summary><i>Previously, woju (Wojtek Porczyk) wrote…</i></summary><blockquote>

If you have all distfiles cached meson won't try to download, so you may be lucky. The mistake is that user *both* fails to unshare *and* forgets to download one or more distfiles. The problem is that, you may try it once and it works (won't download anything), so you do it more and get complacent. Then we update submodules, user pulls and oops, s/he just connected to the network.
</blockquote></details>

Ok, I won't insist on rewording.

**Review by woju — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1018)* status: 0 of 1 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @mkow)

___
*[`Documentation/devel/building.rst` line 324 at r1](https://reviewable.io/reviews/gramineproject/gramine/1018#-NFoCdEE4kFeJXGOnXP5:-NH_HSHE5JMKeUyKn970:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/0921338caa2e28e321cc5851e6c820e3730d9015/Documentation/devel/building.rst#L324)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

The names still don't match for me (e.g. `mbedtls-mbedtls-3.2.1.tar.gz` vs `mbedtls-3.2.1.tar.gz`). Also, even it they matched, we have to describe this, as this option is disabled by default in `wget` (which most likely be used by the users to do this step).
</blockquote></details>

Done.

**Review by woju — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1018)* status: 0 of 1 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @mkow)

___
*[`Documentation/devel/building.rst` line 333 at r3](https://reviewable.io/reviews/gramineproject/gramine/1018#-NHy_8ND5kqm4heOtGWP:-NHy_8ND5kqm4heOtGWQ:bq5kznb) ([raw file](https://github.com/gramineproject/gramine/blob/f6df7c04878928efda102debf73da313577be981/Documentation/devel/building.rst#L333)):*
> ```reStructuredText
> also git submodules. For this you need to create a |~| dummy builddir using
> :command:`meson setup` command (the chosen options don't matter as long as their
> combination is valid).
> ```

The statement "the chosen options don't matter as long as their combination is valid" is not true since static linking of sgx tooling (#773).
 I'll revise this after merging #512, where `makedist.sh` script will appear. It will contain a canonical set of options, which would be good for anybody.

**Review by mkow — CHANGES_REQUESTED** [review]
> Reviewed 1 of 1 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1018)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @woju)

___
*[`Documentation/devel/building.rst` line 333 at r3](https://reviewable.io/reviews/gramineproject/gramine/1018#-NHy_8ND5kqm4heOtGWP:-NIiqof74XRpIAZmkvMw:btcedfy) ([raw file](https://github.com/gramineproject/gramine/blob/f6df7c04878928efda102debf73da313577be981/Documentation/devel/building.rst#L333)):*
<details><summary><i>Previously, woju (Wojtek Porczyk) wrote…</i></summary><blockquote>

The statement "the chosen options don't matter as long as their combination is valid" is not true since static linking of sgx tooling (#773).
 I'll revise this after merging #512, where `makedist.sh` script will appear. It will contain a canonical set of options, which would be good for anybody.
</blockquote></details>

#512 was merged, please update.

**Review by mkow — APPROVED** [review]
> Reviewed 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1018)* status: all files reviewed, all discussions resolved, not enough approvals from different teams (1 more required, approved so far: ITL)

**Review by dimakuv — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/1018)* status: :shipit: complete! all files reviewed, all discussions resolved
